cast/src/lib.rs

```diff
@@ -491,6 +491,64 @@ where
         Ok(self.provider.get_transaction_count(who, block).await?)
     }
 
+    /// # Example
+    ///
+    /// ```no_run
+    /// use cast::Cast;
+    /// use ethers_providers::{Provider, Http};
+    /// use ethers_core::types::Address;
+    /// use std::{str::FromStr, convert::TryFrom};
+    ///
+    /// # async fn foo() -> eyre::Result<()> {
+    /// let provider = Provider::<Http>::try_from("http://localhost:8545")?;
+    /// let cast = Cast::new(provider);
+    /// let addr = Address::from_str("0x7eD52863829AB99354F3a0503A622e82AcD5F7d3")?;
+    /// let implementation = cast.implementation(addr, None).await?;
+    /// println!("{}", implementation);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn implementation<T: Into<NameOrAddress> + Send + Sync>(
+        &self,
+        who: T,
+        block: Option<BlockId>,
+    ) -> Result<String> {
+        let slot =
+            H256::from_str("0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc")?;
+        let value = self.provider.get_storage_at(who, slot, block).await?;
```

Comment:
Should this call a contract's `function implementation() returns (address)` method first, then query the EIP-1967 slot if that call fails? That way it supports non-standard proxies too that don't follow EIP-1967, with compound and compound governors being the common ones

Reply:
Calling `implementation()` second, if the 1967 slot is empty, might be better actually

Reply:
Definitely in favor if this in a follow up PR

Reply:
There's also an `implementation` field in the etherscan `contracts` route: https://docs.rs/ethers/latest/ethers/etherscan/contract/struct.Metadata.html